tutormfe/patches/openedx-lms-production-settings

```diff
@@ -14,3 +14,42 @@ PROFILE_MICROFRONTEND_URL = "{% if ENABLE_HTTPS %}https://{% else %}http://{% en
 LOGIN_REDIRECT_WHITELIST.append("{{ MFE_HOST }}")
 CORS_ORIGIN_WHITELIST.append("{% if ENABLE_HTTPS %}https://{% else %}http://{% endif %}{{ MFE_HOST }}")
 CSRF_TRUSTED_ORIGINS.append("{{ MFE_HOST }}")
+
+# Starts of Dynamic config API settings
+
+ENABLE_MFE_CONFIG_API = True
+
+ALLOWED_HOSTS.append("{{ MFE_HOST }}")
+
+MFE_CONFIG = {
```

Comment:
Wait, why do we still need `templates/mfe/build/mfe/env/production`? I thought that production settings were no longer stored in the MFE?

Reply:
Yeah this PR is not intened to be mergeable yet, I though at somepoint I would add a conditional value for all those changed is they don't want to use config api.  I don't have a strong opinion about that though. 

Secondaly I assume in all cases we are still supporting to build the mfe image locally and if yes, do we load the configuration from where, i.e. where would be the source of congiuration, worth noting we have 4 ways for congiraution
- Production/Development mode General
- Production/Development mode per MFE
And lastly what about this in the equation https://github.com/overhangio/tutor-mfe/pull/62

Reply:
@arbrandes , based on the conversation above @regisb is in favor of making it so tutor-mfe (and thus Olive) only provides runtime MFE configuration. In other words, we wouldn't provide a toggle to fall back to static config.

I am good either way, and @ghassanmas says he has no strong opinion. I am coming to like Régis's proposal because it simplifies tutor-mfe and it is in line with our requirement that all released MFEs must support the dynamic config API. I would like to follow it up with a change to edx-platform which turns on the Runtime Config API by default (2U could still disable it if they wanted to) but that wouldn't be a blocker for Olive.

As the PR maintainer can I get your 👍🏻 on this?

Reply:
In any case  I assume people can still build their mfe image,  and in case we set to dymianc config only, and configuration is saved in lms setting  then  from where  where should the configuration be loaded when Docke is buildig the image? You might argue that it shouldn't care about the config because it would be loaded at runtime, then this open two important question: _Are all configuration varaible runtimeable?_ 
Take the case for favicon.ico edxNuxt made a PR for every MFE to make than config runtimeable,

Reply:
> we don't need to feed Docker with anything

I mean to say that we probably don't need `env/production` overrides except `APP_ID`.

Reply:
> announce that there are breaking changes

In this case, we should consider deprecating static config for this release, then removing it in Palm.

> error/crash at runtime where I think the LMS_BASE/BASE_URL

I admit I did not test the absence of those.  Indeed, I can imagine runtime failing before dynamic config is fetched, though we'd have to see exactly why.  If there's a simple workaround as you suggest (even if a little ugly), we can comment on it and then see about fixing it in `frontend-platform`.

Reply:
Regarding the transition case, I suggest making the patch inside `MFE_CONFIG` but given`MFE_CONFIG` is itself defined inside a patch `openedx-lms-production-settings` and `openedx-lms-development-settings`, then we would be creating a patch inside a ptach... It's just reslicantion not that I have opinoin. 

Secondaly can we ditch the idea of having a patch, and then ask people when they want to overide a variable then they use same patch we used `openedx-lms-production-settings` and `openedx-lms-development-settings`, and it can be something like: 

`MFE_CONFIG["SITE_NAME"]= "something"`
and
`MFE_CONFIG_OVERRIDES["profile"]["SITE_NAME"]="Something else"`

**This would work if we are sure plugin patche will always be loaed before user defined pathces**...  It wouldn't matter if they are creating new config only in case they are overriding. 

So my suggesion above is consdired okay, then I would move with not creating new patches, if it's not okay then I would move with creating new patches i.e. `mfe-dynamic-production` and `mfe-dynamic-development`, then when say in upgrade instruction please move your patches to and use the new format which is json/dict...

Reply:
I don't think we should be creating new patches.  IMHO there are two options:

1. Tell operators to migrate their existing patches, if any, to the database.  This has to be done as part of the upgrade to Olive, and must be high up in the list of actions to take in the Release Notes.
2.  Tell operators about the above, but they have until Palm to do it.  In the meantime, we continue to support static configuration exactly as it was before: this is the gist of Kyle's PR.

@kdmccormick, @regisb, thoughts?

Reply:
Thanks for your detailed analysis @arbrandes . I am OK with the breaking change if you and @regisb are. That said I am a little nervous about making this breaking change three weeks before the release... I am not against it, I just think we should test it very thoroughly, and if there's any indication by Dec 9 that runtime config could introduce regressions, then we should hold of on it until Palm.

Side note:

> Tell operators/plugin authors to migrate their existing patches, if any, to the database. This has to be done as part of the upgrade to Olive, and must be high up in the list of actions to take in the Release Notes.

It'd be unfortunate if this PR forced folks to move their config to the database. My PR allowed the use of a [new tutor-mfe setting](https://github.com/kdmccormick/tutor-mfe/commit/6f6b06d6bc12ac512ff043abf44e7fdeff5e1115#diff-00603dd4ea78f0d1797c3c77634aeefdb1ec967616345f9ff71bf7053e927a4cR18) (`MFE_ENV_COMMON`) to set MFE_CONFIG across apps. If we don't go with that, we should allow & encourage operators to use the `openedx-lms-*-settings` patches before shoving all their config in a db. (I know that db-based config is needed for eduNEXT's use case, but for the basic instance operator, encouraging config-as-code seems more in line with Tutor's philosophy).

Reply:
Wait let me backtrack a little. 

> migrate their existing patches, if any, to the database

What does this sentence actually mean? What is the "database"? Do you mean MySQL? MFE settings are stored in MySQL???

I really hope this is not the case. Moving platform-dependent values to persistent storage makes it very difficult to 1) keep track of changes 2) have different values for different environments (e.g: dev and local). My experience with Ecommerce settings is that it's a bloody disaster.

Do we have any other way to add custom MFE settings to the LMS? For instance, in Django settings? If yes, then we could patch LMS settings in this plugin to add custom MFE configuration to the LMS.

> if there's any indication by Dec 9 that runtime config could introduce regressions, then we should hold of on it until Palm.

We should strive to get runtime config in Tutor in time for Olive. If not, users will face prohibitive image build delays because of the increasing number of MFEs.

Reply:
I just opened an issue #86 regarding config, those are two issue I found so far. 
Regarding where to put the config, I am still not sure if we stick to lms-settigns, how the config would patched, using the current patches (i.e. move them from .env to lms-* thus mean patching a patch inside a patch. Second option seem to follow @kdmccormick suggestion, the only caveate I have with, is how would change spesfic variable without overrwriting the other `tutor config save --set MFE_ENV_COMMON=<do everything need to go here to cahnge spesfic one?>`

Reply:
@ghassanmas Couldn't users override MFE config by patching `openedx-lms-*-settings` like this?
```python
MFE_CONFIG["BASE_URL"] = ...
MFE_CONFIG["SITE_NAME"] = ...
# etc.

MFE_CONFIG_OVERRIDES["learning"]["SOME_SETTING"] = ...
MFE_CONFIG_OVERRIDES["learning"]["ANOTHER_SETTING"] = ...
MFE_CONFIG_OVERRIDES["account"]["YET_ANOTHER_SETTING"] = ...
# etc.
```

or, equivalently:
```python
MFE_CONFIG.update({
    "BASE_URL": ...,
    "SITE_NAME": ...,
    # etc.
})
MFE_CONFIG_OVERRIDES["learning"].update({
    "SOME_SETTING": ...,
    "ANOTHER_SETTING": ...,
})
MFE_CONFIG_OVERRIDES["account"].update({
    "YET_ANOTHER_SETTING": ...,
})
# etc.
```

Reply:
@regisb I am going to do my best to answer you, but please correct me if I'm wrong @arbrandes!

> What does this sentence actually mean? What is the "database"? Do you mean MySQL? MFE settings are stored in MySQL???

The LMS MFE Config API first looks at the `site_configuration` database table, and then falls back to Django settings. This is a pattern used widely throughout edx-platform. It originally existed support edX.org microsites (which are now gone) but some members of the community still rely on it for multi-tenancy. You can read about it [here](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/sites/index.html).

> I really hope this is not the case. Moving platform-dependent values to persistent storage makes it very difficult to 1) keep track of changes 2) have different values for different environments (e.g: dev and local). My experience with Ecommerce settings is that it's a bloody disaster.

I'm with you. I hate `site_configuration`, and while we're on the topic, I hate waffle flags too. [site_configuration is proposed for deprecation](https://github.com/openedx/platform-roadmap/issues/21) but nobody has had the time and/or nerves to carry it out. I encourage you to share you opinion on that issue.... the more voices we have there the easier it'll be to get funding/prioritization.

> Do we have any other way to add custom MFE settings to the LMS? For instance, in Django settings? If yes, then we could patch LMS settings in this plugin to add custom MFE configuration to the LMS.

Yes, Django settings work fine for the MFE Config API. If you don't use `site_configuration`, you can pretend it doesn't exist. When documenting the Config API in this plugin, I think we should omit the `site_configuration` option and just describe the Django settings override option.

Reply:
> Yes I though about that and my concern was if their patch were loaded before tutor-mfe patch, then tutor-mfe might override their overide...

@ghassanmas ah, you are totally right. This will work great if the tutor-mfe patch is loaded first, but it will break if the user patch is loaded first.

R

Reply:
Oh! I think we can solve this by using the `priority` filter callback argument, which Régis literally added to Tutor's hooks API today 😄 (as part of the [jobs PR](https://github.com/overhangio/tutor/pull/738))

By changing [this line from tutor-mfe's plugin.py](https://github.com/overhangio/tutor-mfe/blob/master/tutormfe/plugin.py#L125) from:
```python
        tutor_hooks.Filters.ENV_PATCHES.add_item((os.path.basename(path), patch_file.read()))
```
to:
```python
        tutor_hooks.Filters.ENV_PATCHES.add_item((os.path.basename(path), patch_file.read()), priority=1)
```
we would ensure that tutor-mfe's patches are applied before any user patches (unless the plugin user sets the priority to something higher than 1, in which case they should know what they're doing).

Reply:
That is great! 
![giphy](https://user-images.githubusercontent.com/16361375/202157054-e09ed24d-b40c-4077-a29f-239b039e9f2e.gif)

> we would ensure that tutor-mfe's patches are applied before any user patches (unless the plugin user sets the priority to something higher than 1, in which case they should know what they're doing). 

So this need to be mentioned in docs, so I guess the the higher the priority, the earliest the patch would load... but in our context its a bit confusing because the lower the priority, the more effective the patch would be, given the last to load is the one with lowest priority but it's the one which its configuration would lead...   

So I suppose I need to install tutor from master to try this

Reply:
So in summary the situatoni of this PR _To my end_ 
1. We announce breaking changes for old pathes and then ask people to use the typical patch of llms/cms-prdo/dev/-settings patch: _We are sure this will go smoot beacause by default their patch will have prirority 10, and ours has 1, this it will load earlier. 
1.  Regarding the configuration that need to loads at build time, I created issue #86 
1. Lastly, we are set that tutor-mfe is going to go dynamic config only?, _of course people might need rebuild the image, but rebudiling the image for changing most of the values doesn't affect them_ i.e. you rebuild to override npm pacakge, apply your fork for the MFE...etc 

1. My last concern is for point 3, is that MFEs will take around **0.5 second more for their first load**, _need to testing to ensure its accuarte_, add to the that, how common opreator would need to rebuild their mfe-image to apply npm level changes thus I still see a value for disabling MFEs config. Though not sure the benefit  of the added complexity is a fair tradeoff. Again I don't have strong feeling,  given also MFEs arch might change  openedx/open-edx-proposals/pull/410

Reply:
> MFE settings are stored in MySQL???

I think @kdmccormick took care of that one. :)

It's a pretty important feature of the dynamic config API, though, no matter how unpalatable in terms of config management.  As Kyle also points out, we'd need a (presumably Django settings-based) replacement for per-domain configuration.

Otherwise, I'm fine with the breaking change and the approach to patching.

Reply:
@ghassanmas a lower numerical value actually refers to a "higher priority" in the API, so we are saying that tutor-mfe's patch callback should be executed first.

Reply:
> tutor-mfe is going to go dynamic config only?

On this: we can't remove `MFE_CONFIG_API_URL`, and that needs to be set per MFE.  While we could technically remove `env/development` (because `env/production` is sourced first), I don't think we can remove `env/production`.

Why?  Even if we patch that into Django settings, that configuration will never get fetched... because it's missing the right URL.

An alternative would be to blind-patch each MFE's `.env*` files, which I can't say I'm a fan of.  Or to hard-code `MFE_CONFIG_API_URL` into the Dockerfile build, which I'm even less of a fan of.

Reply:
Alright, at least the API URL would be tutor-configurable in the usual way.

Let's double-check what we're deciding, again:

1. MFE_CONFIG_API_URL will be configurable via `tutor config save`, and set like `APP_ID` in the Dockerfile.
2. `env/*` files will be removed in this plugin.
3. default env settings [in `plugins.py`](https://github.com/overhangio/tutor-mfe/blob/master/tutormfe/plugin.py#L20-L25) will be removed.
4. Default settings (such as the ones in 3 above.) will be ported to `MFE_CONFIG` or `MFE_CONFIG_OVERRIDES` in Django settings via `openedx-[lms|cms]-[production|development]-settings` in this plugin.
5. Downstream plugin authors can override `MFE_CONFIG*` by patching.

My question is: how will a user change `MFE_CONFIG` or `MFE_CONFIG_OVERRIDES` without either having to write a plugin or using the database?  We'd have to add them to [the defaults](https://github.com/overhangio/tutor-mfe/blob/master/tutormfe/plugin.py#L10) and template them out in `openedx-[lms|cms]-[production|development]-settings` in here, correct?

Reply:
@arbrandes 1-3 look good to me 👍🏻 

In terms of 4-5, I see two ways forward:

### (a) Require a plugin in order to modify MFE config

This would be mostly in line with how Tutor LMS/CMS currently works: only a [handful of Django settings are configurable](https://docs.tutor.overhang.io/configuration.html) via Tutor config; for the rest, users should make a plugin to patch the lms/cms settings files.

If I understand @regisb 's philosophy right, it is preferable to keep the Tutor config interface small & easy to comprehend. Complexity like Django settings and MFE settings should be pushed out to the plugin layer. If making a plugin is too big of a lift for Django/MFE settings, then we should handle that by making it easier to write plugins.

### (b) Allow MFE config to be specified via Tutor config

That all said, it'd be pretty easy to add some overridable settings to Tutor MFE. In my original PR I had something like `MFE_CONFIG_COMMON` and `MFE_CONFIG_APPS`, which I used to patch the LMS settings files from tutor-mfe. Worked great.

Reply:
@regisb, can we get a ruling on a) vs b) above?  I'm fine either way, but you know the Tutor audience better than I do.

Reply:
I vote for (b). The initial implementation of the tutor-mfe plugin made it possible to override MFE-specific settings via the Tutor config, but this means that the configuration becomes harder to maintain. It really goes against the whole "you-want-to-customize-openedx-write-a-plugin" philosophy. I realize it's a breaking change, but I'd rather make it sooner than later. I'm not even sure we should add the `MFE_CONFIG_COMMON` and `MFE_CONFIG_APPS` settings.

Reply:
Do you mean you vote for (a) @regisb ?

Reply:
One other argument for (b) that I realized:

The MFE Config API is currently part of LMS. I believe there was a desire to keep it loosely coupled to LMS so that it could be extracted into its own service at some point.

If tutor-mfe users need to patch LMS settings in order to configure MFEs, then extracting the MFE Config API from LMS config would be a breaking change for tutor-mfe users. But if Tutor config is used instead, then the LMS<->MFEConfig coupling could remain an implementation detail.

Reply:
> Do you mean you vote for (a) @regisb ?

God... yes I do... I'll amend my comment.

Reply:
So recap we are discussing between (a) and (b), tutor vison and best practice is aligned with  (a), while the on of advantge of (b) is that opreator can change config at  build time. 

In any case we need to define/set some configuration at build time, and the de-facto of how these configuraiotion come might not be aligned with tutor vision, those are: 

- `APP_ID`, `PUBLIC_PATH`  those are always equals each others and are set through `{{ app["name"] }}` 
- `MFE_CONFIG_API`: we now set this, and we suppose to create MFE config for it; which bring question how would that align with our current proxy, i.e. if we user define it as absoulte url, then the proxy role added would be of no use. 
- `ACCESS_TOKEN_COOKIE_NAME` We need to set it at build time hence #88, but shall it configurabale like `MFE_CONFIG_API` 
- Others configuration @arbrandes pointed above. 

So shall we consdier configuratoin at build time as "ad-hoc basis", or should we have a pattern for them. 

On the bigger picture assuming the vision is to be dynamic config only, I hope at least we can agree that dynamic config as its, still not prefect or incomplete even if **it works** this due to: 

- There will always be latencey for the first time MFE load
- There will be a couple of console errors, due to browser trying resources which its defaults are incorrect, i.e favicon, logo...etc. 
- It's hard to predict which configuration might be updated, hence #88 the case of course auhtoring.

Reply:
@ghassanmas:

`APP_ID` and `PUBLIC_PATH`: sounds right.  They're only set at build time.

`MFE_CONFIG_API_URL`: we should set it at build time via the Dockerfile, like `APP_ID`, pointing it to the Caddy proxy.  Contrary to what I said before, we won't need Tutor config for it.

`ACCESS_TOKEN_COOKIE_NAME`: It doesn't look like the MFE config API endpoint in t he LMS requires authentication, so we might not need to set it at build time.

I think we're in agreement about the drawbacks to the default configuration.  The console errors should be fixed upstream, if we think they're critical enough.

Let's discuss the webpack production config on that PR.  (Thanks for it, btw!)

Reply:
I meant to link this issue #86, regarding `ACCESS_TOKEN_COOKIE_NAME` it's needed not because the config api requries config, its due to what I think is _The frontend-platform Auth service not waiting for dynamic config to define it or soemthing_.

Reply:
Ah, ok.  Then we should set `ACCESS_TOKEN_COOKIE_NAME` directly in the Dockerfile.

Reply:
Thinkking about it now, and hence your comment https://github.com/overhangio/tutor-mfe/pull/69#discussion_r1029424714 , I don't rememeber if or when I was testing I was sure it was returend. But anyway I will test again and if its needed I will define it Dockefile along the `LMS_BASE`